app/module/drivers/sensor/battery/battery_common.c

```diff
@@ -29,15 +29,37 @@ int battery_channel_get(const struct battery_value *value, enum sensor_channel c
     return 0;
 }
 
-uint8_t lithium_ion_mv_to_pct(int16_t bat_mv) {
-    // Simple linear approximation of a battery based off adafruit's discharge graph:
-    // https://learn.adafruit.com/li-ion-and-lipoly-batteries/voltages
+uint8_t lithium_ion_mv_to_pct(int16_t batt_mv) {
+    // Lookup table of slope formulas for calculating remaining battery capacity.
+    // The original values used to calculate slopes come from:
+    //
+    //       https://blog.ampow.com/lipo-voltage-chart/
+    //
+    struct lookup_point {
+        float millivolts;
+        float percent;
+    };
 
-    if (bat_mv >= 4200) {
+    static const struct lookup_point battery_lookup[] = {
+        {.millivolts = 4200, .percent = 100}, {.millivolts = 3870, .percent = 60},
+        {.millivolts = 3690, .percent = 10},  {.millivolts = 3610, .percent = 5},
+        {.millivolts = 3270, .percent = 0},
+    };
+
+    if (batt_mv > 4200) {
         return 100;
-    } else if (bat_mv <= 3450) {
-        return 0;
     }
 
-    return bat_mv * 2 / 15 - 459;
-}
\ No newline at end of file
+    for (int i = 1; i < ARRAY_SIZE(battery_lookup); i++) {
+        struct lookup_point one = battery_lookup[i - 1];
+        struct lookup_point two = battery_lookup[i];
+        if (batt_mv >= two.millivolts) {
+            float slope = (one.percent - two.percent) / (one.millivolts - two.millivolts);
+            float offset = one.percent - slope * one.millivolts;
+            float batt_percent = (slope * batt_mv) + offset;
+            return (uint8_t)batt_percent;
+        }
+    }
+
+    return 0;
```

Comment:
Optional: If this were a general linear interpolation routine, this should be

```c
return battery_lookup[ARRAY_SIZE(battery_lookup) - 1].percent;
```

but since the last entry should always be 0%, this is fine.